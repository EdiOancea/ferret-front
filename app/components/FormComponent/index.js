import React from 'react';
import PropTypes from 'prop-types';
import withStyles from '@material-ui/core/styles/withStyles';
import { Formik, Form } from 'formik';

import Input from 'components/Input';
import FormCheckbox from 'components/FormCheckbox';
import FormButton from 'components/FormButton';

const styles = theme => ({
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing.unit,
  },
});

const renderFields = (fields, options) =>
  fields.map(field => {
    switch (field.type) {
      case 'input':
        return (
          <Input
            key={field.options.name}
            setFieldValue={options.setFieldValue}
            {...field.options}
          />
        );
      case 'button':
        return <FormButton key={field.options.name} {...field.options} />;
      case 'checkbox':
        return <FormCheckbox key={field.options.name} {...field.options} />;
      default:
        return null;
    }
  });

const FormComponent = props => {
  const {
    fields,
    initialValues,
    validationSchema,
    classes,
    onSubmit,
    errorMessage,
  } = props;

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      validationSchema={validationSchema}
      render={({ setFieldValue }) => (
        <Form className={classes.form}>
          {renderFields(fields, { setFieldValue })}
          <div>{errorMessage}</div>
        </Form>
      )}
    />
  );
};

FormComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  fields: PropTypes.array.isRequired,
  initialValues: PropTypes.object.isRequired,
  validationSchema: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
  errorMessage: PropTypes.string.isRequired,
};

export default withStyles(styles)(FormComponent);
