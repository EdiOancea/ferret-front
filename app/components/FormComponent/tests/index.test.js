import React from 'react';
import { render } from 'enzyme';
import FormComponent from 'components/FormComponent';
import * as Yup from 'yup';

describe('<FormComponent />', () => {
  const config = {
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema: Yup.object().shape({
      email: Yup.string()
        .email('Invalid email address.')
        .required('This field is required.'),
      password: Yup.string().required('This field is required'),
    }),
    fields: [
      {
        type: 'input',
        options: {
          type: 'email',
          label: 'Email Address',
          name: 'email',
        },
      },
      {
        type: 'input',
        options: {
          type: 'password',
          label: 'Password',
          name: 'password',
        },
      },
      {
        type: 'checkbox',
        options: {
          label: 'Remember me',
          name: 'checkbox',
        },
      },
      {
        type: 'button',
        options: {
          text: 'SIGN IN',
          name: 'submitButton',
        },
      },
      {
        type: 'wrongtype',
        options: {},
      },
    ],
  };

  it('Should not regress', () => {
    const tree = render(
      <FormComponent
        {...config}
        onSubmit={() => null}
        errorMessage="test error message"
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
