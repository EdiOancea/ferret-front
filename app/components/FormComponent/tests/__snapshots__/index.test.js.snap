// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormComponent /> Should not regress 1`] = `
<form
  class="FormComponent-form-1"
>
  <div>
    <div
      class="MuiFormControl-root-2 MuiFormControl-fullWidth-5"
    >
      <label
        class="MuiFormLabel-root-17 MuiInputLabel-root-6 MuiInputLabel-formControl-11 MuiInputLabel-animated-14"
        data-shrink="false"
      >
        Email Address
      </label>
      <div
        class="MuiInputBase-root-37 MuiInput-root-24 MuiInput-underline-28 MuiInputBase-fullWidth-46 MuiInput-fullWidth-31 MuiInputBase-formControl-38 MuiInput-formControl-25"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input-47 MuiInput-input-32 MuiInputBase-inputType-50 MuiInput-inputType-35"
          name="email"
          type="email"
        />
      </div>
    </div>
  </div>
  <div>
    <div
      class="MuiFormControl-root-2 MuiFormControl-fullWidth-5"
    >
      <label
        class="MuiFormLabel-root-17 MuiInputLabel-root-6 MuiInputLabel-formControl-11 MuiInputLabel-animated-14"
        data-shrink="false"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root-37 MuiInput-root-24 MuiInput-underline-28 MuiInputBase-fullWidth-46 MuiInput-fullWidth-31 MuiInputBase-formControl-38 MuiInput-formControl-25"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input-47 MuiInput-input-32 MuiInputBase-inputType-50 MuiInput-inputType-35"
          name="password"
          type="password"
        />
      </div>
    </div>
  </div>
  <label
    class="MuiFormControlLabel-root-54"
  >
    <span
      class="MuiButtonBase-root-76 MuiIconButton-root-70 MuiPrivateSwitchBase-root-66 MuiCheckbox-root-60 MuiCheckbox-colorPrimary-64"
    >
      <span
        class="MuiIconButton-label-75"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-79"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <input
          class="MuiPrivateSwitchBase-input-69"
          data-indeterminate="false"
          type="checkbox"
          value="remember"
        />
      </span>
    </span>
    <span
      class="MuiTypography-root-88 MuiTypography-body1-97 MuiFormControlLabel-label-59"
    >
      Remember me
    </span>
  </label>
  <button
    class="MuiButtonBase-root-76 MuiButton-root-125 MuiButton-contained-136 MuiButton-containedPrimary-137 MuiButton-raised-139 MuiButton-raisedPrimary-140 MuiButton-fullWidth-150 FormButton-submit-124"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label-126"
    >
      SIGN IN
    </span>
  </button>
  <div>
    test error message
  </div>
</form>
`;
